File: examples/swift/filtering.swift
import Foundation

@main
struct FilteringExample {
    static func main() {
        do {
            try run()
        } catch {
            fatalError("Swift filtering example failed: \(error)")
        }
    }

    static func run() throws {
        var reader = try reader_from_bytes(decodeSample())

        print("Messages in time range [0, 10]:")
        var stream = try reader_messages(reader)
        try message_stream_time_range(stream, 0, 10)
        var count = 0
        while true {
            let item = try message_stream_next(stream)
            if !item.has_value {
                break
            }
            let msg = item.value
            print("  time=\(msg.log_time) channel=\(msg.channel_id)")
            count += 1
            if count >= 5 {
                break
            }
        }
        reader = try message_stream_into_reader(stream)

        print("\nMessages on channels [1]:")
        stream = try reader_messages(reader)
        count = 0
        while true {
            let item = try message_stream_next(stream)
            if !item.has_value {
                break
            }
            let msg = item.value
            if msg.channel_id != 1 {
                continue
            }
            print("  time=\(msg.log_time) channel=\(msg.channel_id)")
            count += 1
            if count >= 5 {
                break
            }
        }
        reader = try message_stream_into_reader(stream)

        print("\nMessages on channel 1 in time range [0, 10]:")
        stream = try reader_messages(reader)
        count = 0
        while true {
            let item = try message_stream_next(stream)
            if !item.has_value {
                break
            }
            let msg = item.value
            if msg.channel_id != 1 || msg.log_time > 10 {
                continue
            }
            print("  time=\(msg.log_time) channel=\(msg.channel_id)")
            count += 1
            if count >= 5 {
                break
            }
        }
    }
}

File: examples/cpp/filtering.cc
#include "common.h"
#include "cxx.h"
#include "mcapable-cpp/src/lib.rs.h"
#include <iostream>

int main() {
  auto data = sample_bytes();
  auto reader = mcapable::reader_from_bytes(data);

  auto stream = mcapable::reader_messages(*reader);

  std::cout << "Messages in time range [0, 10]:\n";
  mcapable::Message message{};
  size_t count = 0;
  while (mcapable::message_stream_next(*stream, message)) {
    if (message.log_time < 0 || message.log_time > 10) {
      continue;
    }
    std::cout << "  time=" << message.log_time << " channel=" << message.channel_id
              << "\n";
    if (++count >= 5) {
      break;
    }
  }
  reader = mcapable::message_stream_into_reader(*stream);

  stream = mcapable::reader_messages(*reader);

  std::cout << "\nMessages on channels [1]:\n";
  count = 0;
  while (mcapable::message_stream_next(*stream, message)) {
    if (message.channel_id != 1) {
      continue;
    }
    std::cout << "  time=" << message.log_time << " channel=" << message.channel_id
              << "\n";
    if (++count >= 5) {
      break;
    }
  }
  reader = mcapable::message_stream_into_reader(*stream);

  stream = mcapable::reader_messages(*reader);

  std::cout << "\nMessages on channel 1 in time range [0, 10]:\n";
  count = 0;
  while (mcapable::message_stream_next(*stream, message)) {
    if (message.channel_id != 1 || message.log_time > 10) {
      continue;
    }
    std::cout << "  time=" << message.log_time << " channel=" << message.channel_id
              << "\n";
    if (++count >= 5) {
      break;
    }
  }
  reader = mcapable::message_stream_into_reader(*stream);

  return 0;
}
